src/DataTypes/Serializations/SerializationMap.cpp

```diff
@@ -140,14 +143,27 @@ void SerializationMap::deserializeTextImpl(IColumn & column, ReadBuffer & istr,
             if (*istr.position() == '}')
                 break;
 
-            reader(key, key_column);
-            skipWhitespaceIfAny(istr);
-            assertChar(':', istr);
+            /// For int keys replace ':' to space, because serialization of numbers
+            /// uses unsafe version of reading from text, which consumes ':' symbol
+            /// and has UB in that case.
+            if (is_integer_key)
+            {
+                auto * colon_pos = find_first_symbols<':'>(istr.position(), istr.buffer().end());
```

Comment:
Seems like it does not completely fixes the issue, because it may not find a colon if current position is near to the end of the buffer and number intersects the border. It means that parsing will not work for large inputs. However, I don't know any good way to fix it in general case, so current solution is possibly ok (but at least we should add a comment about it).